authz/providers/azure/azure.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright The Guard Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package azure
+
+import (
+	"strings"
+
+	auth "github.com/appscode/guard/auth/providers/azure"
+	"github.com/appscode/guard/authz"
+	"github.com/appscode/guard/authz/providers/azure/rbac"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	authzv1 "k8s.io/api/authorization/v1"
+)
+
+const (
+	OrgType = "azure"
+)
+
+func init() {
+	authz.SupportedOrgs = append(authz.SupportedOrgs, OrgType)
+}
+
+type Authorizer struct {
+	rbacClient *rbac.AccessInfo
+}
+
+type authzInfo struct {
+	AADEndpoint string
+	ARMEndPoint string
+}
+
+func New(opts Options, authopts auth.Options, dataStore authz.Store) (authz.Interface, error) {
+	c := &Authorizer{}
+
+	authzInfoVal, err := getAuthzInfo(authopts.Environment)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error in getAuthzInfo %s")
+	}
+
+	switch opts.AuthzMode {
+	case ARCAuthzMode:
+		c.rbacClient, err = rbac.New(authopts.ClientID, authopts.ClientSecret, authopts.TenantID, authzInfoVal.AADEndpoint, authzInfoVal.ARMEndPoint, opts.AuthzMode, opts.ResourceId, opts.ARMCallLimit, dataStore, opts.SkipAuthzCheck, opts.AuthzResolveGroupMemberships, opts.SkipAuthzForNonAADUsers)
+	case AKSAuthzMode:
+		c.rbacClient, err = rbac.NewWithAKS(opts.AKSAuthzURL, authopts.TenantID, authzInfoVal.ARMEndPoint, opts.AuthzMode, opts.ResourceId, opts.ARMCallLimit, dataStore, opts.SkipAuthzCheck, opts.AuthzResolveGroupMemberships, opts.SkipAuthzForNonAADUsers)
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create ms rbac client")
+	}
+	return c, nil
+}
+
+func (s Authorizer) Check(request *authzv1.SubjectAccessReviewSpec) (*authzv1.SubjectAccessReviewStatus, error) {
+	if request == nil {
+		return nil, errors.New("subject access review is nil")
+	}
+
+	// check if user is service account
+	if strings.HasPrefix(strings.ToLower(request.User), "system") {
```

Comment:
We should check against prefix with delimeter `system` -> `system:serviceaccount:`

Reply:
Although comment mentions service account, idea here is to return 'no opinion' for all system accounts like system:node, system:kube-scheduler etc along with service accounts. I will update the comment and log statement.

Reply:
Sounds good. Please add the delimiter `:` to make sure it is checked correctly.

`strings.HasPrefix(strings.ToLower(request.User), "system:")`